File: unity/test/Src/Cases/API/ForofTest.cs
using NUnit.Framework;
using System;
using System.Collections.Generic;
using Puerts;

namespace Puerts.UnitTest
{
    [UnityEngine.Scripting.Preserve]
    public struct TestVector
    {
        public double x;
        public double y;
        public double z;

        public TestVector(double x, double y, double z) 
        {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        public override string ToString() 
        { 
            return this.x + " " + this.y + " " + this.z;
        }
    }
    public class ForofTestHelper 
    {
        [UnityEngine.Scripting.Preserve]
        public static List<string> GetAStringList() 
        {
            return new List<string>()
            {
                "puerts",
                "really",
                "good"
            };
        }
    }
    [TestFixture]
    public class ForofTest
    {
        [Test]
        public void ListDictForofTest()
        {
            var env = UnitTestEnv.GetEnv();
            env.ExecuteModule("CSharpModuleTest/forof_test.mjs");
            env.Tick();
        }
        
        [Test]
        public void ValueTypeListTest()
        {
            var jsEnv = UnitTestEnv.GetEnv();
            var list = new TestVector[] 
            {
                new TestVector(4, 6, 1)
            };
            var ret = jsEnv.Eval<Func<TestVector[], string>>(@"
                (function(list) {
                    var vec = list.get_Item(0);
                    return `${vec.x} ${vec.y} ${vec.z}`
                })
            ");
            Assert.AreEqual("4 6 1", ret(list));
            jsEnv.Tick();
        }
    }
}

File: unity/test/Src/Resources/CSharpModuleTest/forof_test.mjs
let Assert_AreEqual = CS.NUnit.Framework.Assert.AreEqual


let IntList = puer.$generic(CS.System.Collections.Generic.List$1, CS.System.Int32);
let list = new IntList();
list.Add(1);
list.Add(2);

let IntDict = puer.$generic(CS.System.Collections.Generic.Dictionary$2, CS.System.Int32, CS.System.Int32);
let dict = new IntDict();
dict.Add(10, 20);
dict.Add(30, 40);
let total = 0;
for (const listElement of list) {
    total += listElement;
}
for (const dictElement of dict) {
    total += dictElement.Key + dictElement.Value;
}
Assert_AreEqual(total, 103);


let StringList = puer.$generic(CS.System.Collections.Generic.List$1, CS.System.String);
let listFromCS = CS.Puerts.UnitTest.ForofTestHelper.GetAStringList();
let ret = [];
for (const element of listFromCS) {
    ret.push(element);
}
Assert_AreEqual(ret.join(" "), "puerts really good");